### PR TITLE
[core] Introduce withReadType in ReadBuilder

### PR DESCRIPTION
**Pull request by Zouxxyy**
### Purpose

To #4209

#### public API

```java
 /**
   * Push row type to the reader, support nested row pruning.
   *
   * @param readType read row type, can be a subset of {@link Table#rowType()}
   * @since 1.0.0
   */
  ReadBuilder withReadType(RowType readType);
```

mark `ReadBuilder withProjection(int[][] projection)` and `ReadBuilder withProjection(int[] projection)` as `Deprecated`

#### how to use
```java
RowType writeType =
        DataTypes.ROW(
                DataTypes.FIELD(0, "pt", DataTypes.INT()),
                DataTypes.FIELD(1, "a", DataTypes.INT()),
                DataTypes.FIELD(2, "f0", DataTypes.INT()),
                DataTypes.FIELD(
                        3,
                        "f1",
                        DataTypes.ROW(
                                DataTypes.FIELD(4, "f0", DataTypes.INT()),
                                DataTypes.FIELD(5, "f1", DataTypes.INT()),
                                DataTypes.FIELD(6, "f2", DataTypes.INT()))));
// write
// GenericRow.of(0, 0, 0, GenericRow.of(10, 11, 12))


RowType readType =
        DataTypes.ROW(
                DataTypes.FIELD(
                        3,
                        "f1",
                        DataTypes.ROW(
                                DataTypes.FIELD(4, "f0", DataTypes.INT()),
                                DataTypes.FIELD(6, "f2", DataTypes.INT()))));

ReadBuilder readBuilder = table.newReadBuilder().withReadType(readType);

// read
// GenericRow.of(GenericRow.of(10, 12))
```

### CODE REVIEW

**Comment by JingsongLi** [comment]
We just need a `pruneColumns(RowType requiredSchema)`.

**Comment by Zouxxyy** [comment]
RowType contains all the information (field name, field id, nested structure ... ), it can replace projection

The final API will be modified to like this

```java
    @Deprecated
    default ReadBuilder withProjection(int[] projection) {
        // projection -> requiredSchema
        return pruneColumns(RowType requiredSchema);
    }

    ReadBuilder pruneColumns(RowType requiredSchema);
```

File: paimon-common/src/main/java/org/apache/paimon/table/SystemFields.java [hunk 1]
```diff
@@ -42,8 +43,12 @@ public class SystemFields {
     public static final DataField LEVEL =
             new DataField(Integer.MAX_VALUE - 3, "_LEVEL", DataTypes.INT().notNull());
 
+    public static final DataField ROW_KIND =
```
Comment (JingsongLi) [inline paimon-common/src/main/java/org/apache/paimon/table/SystemFields.java:46]: Add comment: only used by AuditLogTable

File: paimon-common/src/main/java/org/apache/paimon/types/RowType.java [hunk 2]
```diff
@@ -177,6 +191,25 @@ public boolean equals(Object o) {
         return fields.equals(rowType.fields);
     }
 
+    public boolean subsetOf(Object o) {
```
Comment (JingsongLi) [inline paimon-common/src/main/java/org/apache/paimon/types/RowType.java:194]: override

File: paimon-common/src/main/java/org/apache/paimon/types/DataType.java [hunk 3]
```diff
@@ -145,6 +145,10 @@ public boolean equals(Object o) {
         return isNullable == that.isNullable && typeRoot == that.typeRoot;
     }
 
+    public boolean subsetOf(Object o) {
```
Comment (JingsongLi) [inline paimon-common/src/main/java/org/apache/paimon/types/DataType.java:148]: add comment to explain this method.

File: paimon-common/src/main/java/org/apache/paimon/types/RowType.java [hunk 4]
```diff
@@ -223,6 +256,10 @@ public RowType appendDataField(String name, DataType type) {
         return new RowType(newFields);
     }
 
+    public RowType project(int[][] mapping) {
+        return project(Arrays.stream(mapping).mapToInt(arr -> arr[0]).toArray());
```
Comment (JingsongLi) [inline paimon-common/src/main/java/org/apache/paimon/types/RowType.java:260]: remove this temporary method, and just inline this to the invoker.

File: paimon-common/src/main/java/org/apache/paimon/types/RowType.java [hunk 5]
```diff
@@ -238,6 +275,53 @@ public RowType project(List<String> names) {
                         .collect(Collectors.toList()));
     }
 
+    public RowType project(String... names) {
+        return project(Arrays.asList(names));
+    }
+
+    // These methods are just for compatibility with old code to get the projection, once
```
Comment (JingsongLi) [inline paimon-common/src/main/java/org/apache/paimon/types/RowType.java:282]: If it is just for old code, please not add these methods in public `RowType`.

File: paimon-common/src/main/java/org/apache/paimon/types/RowType.java [hunk 6]
```diff
@@ -238,6 +275,53 @@ public RowType project(List<String> names) {
...
+    // toProjection and buildWithNewProjection are not used anymore, remove them.
+    public RowType copyAndSetOriginalRowType(RowType originalRowType) {
+        if (this.originalRowType != null && !originalRowType.equals(this.originalRowType)) {
+            throw new RuntimeException();
+        }
+        return ((RowType) this.copy()).setOriginalRowType(originalRowType);
+    }
+
+    public int[] toProjection() {
```
Comment (JingsongLi) [inline paimon-common/src/main/java/org/apache/paimon/types/RowType.java:291]: it is hard to understand `RowType.toProjection`

File: paimon-common/src/main/java/org/apache/paimon/types/RowType.java [hunk 7]
```diff
@@ -238,6 +275,53 @@ public RowType project(List<String> names) {
...
+        if (originalRowType == null) {
+            throw new RuntimeException();
+        }
+        return fields.stream()
+                .mapToInt(field -> originalRowType.getFieldIndexByFieldId(field.id()))
+                .toArray();
+    }
+
+    // create new RowType based on the new projection and originalRowType
+    public RowType buildWithNewProjection(int[] projection) {
```
Comment (JingsongLi) [inline paimon-common/src/main/java/org/apache/paimon/types/RowType.java:304]: remove this

File: paimon-common/src/main/java/org/apache/paimon/types/RowType.java [hunk 8]
```diff
@@ -56,6 +56,7 @@ public final class RowType extends DataType {
 
     private final List<DataField> fields;
     private InternalRow.FieldGetter[] fieldGetters;
+    private RowType originalRowType;
```
Comment (JingsongLi) [inline paimon-common/src/main/java/org/apache/paimon/types/RowType.java:59]: it should not be here.

**Comment by Zouxxyy** [comment]
@JingsongLi Thanks for review, updated

File: paimon-core/src/main/java/org/apache/paimon/table/source/ReadBuilder.java [hunk 9]
```diff
@@ -111,25 +112,29 @@ default ReadBuilder withFilter(List<Predicate> predicates) {
...
-     * <p>NOTE: Nested row projection is currently not supported.
+     * @param readType read row type, can be a pruned type from {@link Table#rowType()}
+     * @since 1.0.0
      */
-    default ReadBuilder withProjection(int[] projection) {
+    ReadBuilder withReadType(RowType readType);
+
+    /** Apply projection to the reader, Use {@link #withReadType(RowType)} instead. */
+    @Deprecated
+    ReadBuilder withProjection(int[] projection);
```
Comment (JingsongLi) [inline paimon-core/src/main/java/org/apache/paimon/table/source/ReadBuilder.java:124]: I think this method can always be invalid. Not need to be Deprecated.

**Review by JingsongLi — APPROVED** [review]
+1

**Pull request merged**